src/citrine/builders/design_spaces.py

```diff
@@ -0,0 +1,190 @@
+    description: str
+        description for the EnumeratedDesignSpace
+
+    """
+    # Check that the passed balance component is in the grid keys
+    assert balance_component in list(formulation_grids.keys()), \
+        "balance component must be in formulation_grids' keys"
+    non_balance_comps = [comp for comp in list(formulation_grids.keys())
+                         if comp != balance_component]
+    non_balance_descriptors = [dd for dd in descriptors
+                               if dd.key != balance_component]
+    non_balance_grids = {
+        k: v for k, v in formulation_grids.items()
+        if k in non_balance_comps
+    }
+
+    # Check that all grid values are between 0 and 1
+    for kk, val_list in formulation_grids.items():
+        assert all([x >= 0 for x in val_list]) and \
+            all([x <= 1 for x in val_list]), \
+            'values in {} are not between 0 and 1'.format(kk)
+
+    # Start by making a naive product design space of non-balance components
+
+    form_ds = pd.DataFrame(
+        enumerate_cartesian_product(
+            design_grids=non_balance_grids,
+            descriptors=non_balance_descriptors
+        ).data
+    )
+
+    # Re-calculate the balance component column as 1-all others
+    form_ds[balance_component] = form_ds.apply(
+        lambda r: 1 - sum([r[col] for col in non_balance_comps]),
+        axis=1
+    )
+
+    # Order columns so balance component is first
+    form_ds = form_ds[[balance_component] + non_balance_comps]
+
+    # Eliminate out-of-range rows
+    balance_range = (
+        min(formulation_grids[balance_component]),
+        max(formulation_grids[balance_component])
+    )
+    form_ds = form_ds[form_ds[balance_component] >= balance_range[0]]
+    form_ds = form_ds[form_ds[balance_component] <= balance_range[1]]
+    data = form_ds.to_dict('records')
+    design_space = EnumeratedDesignSpace(name=name,
+                                         description=description,
+                                         descriptors=descriptors,
+                                         data=data)
+    return design_space
+
+
+def cartesian_join_design_spaces(
+    design_space_list: List[EnumeratedDesignSpace],
+    name: str = 'Joined enumerated design space',
+    description: str = '',
+) -> EnumeratedDesignSpace:
+    """[ALPHA] Cartesian join of multiple enumerated design spaces.
+
+    Given a list of multiple input EnumeratedDesignSpaces, create a new
+    EnumeratedDesignSpace that is the Cartesian product of candidates from each
+    input design space.
+
+    Ultimate length will be the product of the number of candidates in each
+    individual design space
+
+    Parameters
+    ----------
+    design_space_list: List[EnumeratedDesignSpace]
+        A list of EnumeratedDesignSpaces
+
+    """
+    assert 'join_key' not in [ds.data[0].keys() for ds in design_space_list], \
+        '"join_key" cannot be a descriptor key when using this function'
+
```

Comment:
What happens if some of the design spaces share descriptor keys?  If that's not allowed, then we should check it and raise a `ValueError` if they do.

Reply:
added